File: RentalModule.Application/CommandHandlers/CarOffer/CreateCarOffer/CreateCarOfferCommandHandler.cs
using System.Collections.Immutable;
using RentalModule.Application.Contract.CarOffers.CreateCarOffer;
using RentalModule.Domain.Models;
using RentalModule.Domain.Models.Ids;
using RentalModule.Domain.RepositoryInterfaces.CarOffer;
using RentalModule.Infrastructure.DataBase.Entities;
using Results.Application;
using Results.Domain;
using Shared.CQRS;
using Shared.CQRS.CommandHandlers;

namespace RentalModule.Application.CommandHandlers.CarOffer.CreateCarOffer
{
    public class CreateCarOfferCommandHandler : ICommandHandler<CreateCarOfferCommand,
        HandlerResult<CreateCarOfferResponse, IErrorResult>>
    {
        private readonly ICarOfferCommandRepository _commandRepository;
        private readonly ICarOfferQueryRepository _queryRepository;

        public CreateCarOfferCommandHandler(
            ICarOfferCommandRepository commandRepository,
            ICarOfferQueryRepository queryRepository)
        {
            _commandRepository = commandRepository;
            _queryRepository = queryRepository;
        }

        public async Task<HandlerResult<CreateCarOfferResponse, IErrorResult>> Handle(CreateCarOfferCommand request,
            CancellationToken cancellationToken)
        {
            var carOffer = await _queryRepository.GetByUrlAsync(request.UrlHandle, cancellationToken);
            if (carOffer is not null)
            {
                return new EntityAlreadyExistsErrorResult
                {
                    Title = "Cannot create car offer",
                    Message = $"Car offer with URL handle {request.UrlHandle} already exists."
                };
            }

            var newCarOfferId = new CarOfferId(Guid.NewGuid());
            var carOfferToAdd = new CarOfferModel
            {
                Id = newCarOfferId,
                Heading = request.Heading,
                ShortDescription = request.ShortDescription,
                FeaturedImageUrl = request.FeaturedImageUrl,
                UrlHandle = request.UrlHandle,
                Horsepower = request.Horsepower,
                YearOfProduction = request.YearOfProduction,
                EngineDetails = request.EngineDetails,
                DriveDetails = request.DriveDetails,
                GearboxDetails = request.GearboxDetails,
                CarDeliverylocation = request.CarDeliverylocation,
                CarReturnLocation = request.CarReturnLocation,
                PublishedDate = request.PublishedDate,
                Visible = request.Visible,
                Tarrif = new CarTariffModel
                {
                    Id = new CarTariffId(new Guid()),
                    CarOfferId = newCarOfferId,
                    OneNormalDayPrice = request.OneNormalDayPrice,
                    OneWeekendDayPrice = request.OneWeekendDayPrice,
                    FullWeekendPrice = request.FullWeekendPrice,
                    OneWeekPrice = request.OneWeekPrice,
                    OneMonthPrice = request.OneMonthPrice
                },
                Tags = ConvertToTagModels(request.Tags,
                    newCarOfferId),
                ImageUrls = ConvertToImageUrlModels(request.ImageUrls,
                    newCarOfferId),
                UnavailableDates = ImmutableArray<TimePeriodModel>.Empty
            };

            var addedCarOffer = await _commandRepository.AddAsync(carOfferToAdd, cancellationToken);

            return new CreateCarOfferResponse()
            {
                Id = addedCarOffer.Id,
                Title = "Car offer created",
                Message = $"Car offer '{addedCarOffer.Heading}' was created successfully."
            };
        }

        private static ImmutableArray<CarTagModel> ConvertToTagModels(ImmutableArray<string>? tags,
            CarOfferId carOfferId)
        {
            if (tags == null || tags.Value.IsEmpty) return ImmutableArray<CarTagModel>.Empty;

            return tags.Value.Select(tag => new CarTagModel
            {
                Id = new CarTagId(Guid.NewGuid()),
                Name = tag,
                CarOfferId = carOfferId
            }).ToImmutableArray();
        }
        
        private static ImmutableArray<ImageUrlModel> ConvertToImageUrlModels(ImmutableArray<string>? urls,
            CarOfferId carOfferId)
        {
            if (urls == null || urls.Value.IsEmpty) return ImmutableArray<ImageUrlModel>.Empty;

            return urls.Value.Select(url => new ImageUrlModel
            {
                Id = new ImageUrlId(Guid.NewGuid()),
                Url = url,
                CarOfferId = carOfferId,

            }).ToImmutableArray();
        }
    }
}

File: RentalApi/Application/command_handlers/car_offer/create_car_offer/create_car_offer_command_handler.go
package commands

import (
	"context"
	"log"
	contract "rental-api/Application.contract/car_offers/create_car_offer"
	"rental-api/Application/services"
	models "rental-api/Domain/models/domestic"
	repository_interfaces "rental-api/Domain/repository_interfaces/car_offer_repository"
	"rental-api/Domain/responses"
	data_fetcher "rental-api/Domain/service_interfaces"
)

type CreateCarOfferCommandHandler struct {
	carOfferCommandRepository repository_interfaces.CarOfferCommandRepository
	dataFetcher               data_fetcher.MicroserviceConnector
}

func NewCreateCarOfferCommandHandler(
	carOfferCommandRepository repository_interfaces.CarOfferCommandRepository,
	dataFetcher data_fetcher.MicroserviceConnector,
) *CreateCarOfferCommandHandler {
	return &CreateCarOfferCommandHandler{
		carOfferCommandRepository: carOfferCommandRepository,
		dataFetcher:               dataFetcher,
	}
}

func (h *CreateCarOfferCommandHandler) Handle(ctx context.Context, command *CreateCarOfferCommand) (*contract.CreateCarOfferResponse, error) {
	userInfo, err := h.dataFetcher.GetUserInternalInfo(command.JwtToken)
	log.Println(userInfo)
	if err != nil {
		response := responses.NewResponse[contract.CreateCarOfferResponse](401, "Unauthorized")
		return &response, nil
	}
	if !services.IsAdminOrSuperAdmin(userInfo.Roles) {
		response := responses.NewResponse[contract.CreateCarOfferResponse](403, "Not authorized to create a car offer")
		return &response, nil
	}
	carOffer := &models.CarOfferModel{
		Heading:            command.Heading,
		ShortDescription:   command.ShortDescription,
		FeaturedImageUrl:   command.FeaturedImageUrl,
		UrlHandle:          command.UrlHandle,
		Horsepower:         command.Horsepower,
		YearOfProduction:   command.YearOfProduction,
		EngineDetails:      command.EngineDetails,
		DriveDetails:       command.DriveDetails,
		GearboxDetails:     command.GearboxDetails,
		Visible:            command.Visible,
		OneNormalDayPrice:  command.OneNormalDayPrice,
		OneWeekendDayPrice: command.OneWeekendDayPrice,
		OneWeekPrice:       command.OneWeekPrice,
		OneMonthPrice:      command.OneMonthPrice,
		CustodianId:        userInfo.ID,
		CustodianEmail:     userInfo.EmailAddress,
	}

	result, err := h.carOfferCommandRepository.CreateCarOffer(ctx, carOffer, command.Tags, command.ImageUrls)
	if err != nil {
		response := responses.NewResponse[contract.CreateCarOfferResponse](500, "Failed to create car offer")
		return &response, nil
	}

	return &contract.CreateCarOfferResponse{
		BaseResponse: responses.NewBaseResponse(200, "Car offer created successfully"),
		Id:           *result,
	}, nil
}
